packages/chatbot-server-mongodb-public/src/tracing/mongoDbChatbotCommentSentiment.ts

```diff
@@ -0,0 +1,149 @@
+  messageWithCommentId: ObjectId;
+}
+
+const Sentiment = z.object({
+  reasoning: z.string({
+    description: "Concise reasoning justifying the sentiment",
+  }),
+  sentiment: z.enum(["positive", "neutral", "negative"]),
+});
+
+export type Sentiment = z.infer<typeof Sentiment>;
+
+const commentSentimentToolDefinition: OpenAI.ChatCompletionTool = {
+  type: "function",
+  function: {
+    name: "comment_sentiment",
+    description: "Calculate the sentiment of a comment in a conversation",
+    parameters: zodToJsonSchema(Sentiment, {
+      $refStrategy: "none",
+    }),
+  },
+};
+
+const systemMessage = {
+  role: "system",
+  content: `You are a quality assurance analyst. Your job is to calculate the sentiment of a comment in a conversation with a chatbot.
+The chatbot you are analyzing answers questions about MongoDB's products and services.
+You are given a snippet of a conversation between a user and the chatbot with some metadata about the conversation, including user ratings on messages and the relevant message with comment.`,
+} satisfies OpenAI.Chat.Completions.ChatCompletionSystemMessageParam;
+
+export const makeJudgeMongoDbChatbotCommentSentiment = (openAiClient: OpenAI) =>
+  wrapTraced(
+    async function ({
+      judgeLlm,
+      messages,
+      messageWithCommentId,
+    }: CommentSentimentParams) {
+      const userMessage = makeUserMessage(messages, messageWithCommentId);
+
+      const messagesForLlm = [systemMessage, userMessage];
+
+      const { sentiment, reasoning } = await getSentiment(
+        openAiClient,
+        judgeLlm,
+        messagesForLlm
+      );
+
+      const SENTIMENT_SCORES = {
+        positive: 1,
+        negative: 0,
+        neutral: 0.5,
+      } as const;
+
+      const score = SENTIMENT_SCORES[sentiment];
+
+      return {
+        name: "CommentSentiment",
+        score,
+        metadata: {
+          reasoning,
+          sentiment,
+        },
+      };
+    },
+    {
+      name: "JudgeMongoDbChatbotCommentSentiment",
+    }
+  );
+
+async function getSentiment(
+  openAiClient: OpenAI,
+  judgeLlm: string,
+  messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[]
+) {
+  const modelRes = await openAiClient.chat.completions.create({
+    messages: messages,
+    model: judgeLlm,
+    tools: [commentSentimentToolDefinition],
+    stream: false,
+    tool_choice: {
+      function: { name: commentSentimentToolDefinition.function.name },
+      type: "function",
+    },
+  });
+  return Sentiment.parse(
+    JSON.parse(
+      modelRes.choices?.[0].message.tool_calls?.[0].function.arguments ?? "{}"
+    )
+  );
+}
+
+function makeUserMessage(
+  messages: Message[],
+  messageWithCommentId: ObjectId
+): OpenAI.Chat.Completions.ChatCompletionUserMessageParam {
+  const commentIdx = messages.findLastIndex(
+    (message) => message.id.equals(messageWithCommentId)
+    // &&
+    //   message.role === "assistant" &&
+    //   message.userComment !== undefined
```

Comment:
Do we need these? Technically correct though perhaps redundant if we already have a message ID that we know has a comment.

Reply:
added a refactored form of the check